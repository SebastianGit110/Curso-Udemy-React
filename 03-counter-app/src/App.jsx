import { Fragment } from "react";
import PropTypes from "prop-types";

// La importacion del componente fragment <Fragment><Fragment /> hace lo mismo que <></> que es envolver codigo jsx para no tener que crear un div ya que en react los componentes deben devolver un único elemento padre.

// <h1> { expr de js menos objs ni funciones async } </h1> react no permite tener hijos obj no funciones async que son obj.
// Para poder imprimir un obj lo tenemos que pasar a JSON con JSON.stringify()

// Se recomienda evitar declarar variables dentro del cuerpo de un componente si su valor no va a cambiar o no está relacionado con un hook. Esto se debe a que React solo re-renderiza la parte del componente que está relacionada con el estado (state) y los props. Ya que cuando el comp se re-renderice va a volver a especificar en memoria el espacio de lo que esta en el cuerpo de la funcion.

// Las props se reciben como si fuera un obj
// En create-react-app ya vienen instaladas las dependencias de propTypes mientras que en vite no
// PropTypes es para definir el tipo a las props o si es obligatorio mandarlas

const nombre = {
  name: "Sebas",
  code: "Sebrat",
};

const funcion = () => "hola";

function App({ title, subtitulo }) {
  // Se pueden poner las prop por defecto en los parametros de la funcion por si no se pasa esa prop se toma ese valor, pero con las defaultProp se puede hacer tambien

  // Se puede hacer asi pero se extiende el cod del comp innecesariamente, para eso estan los PropTypes
  //   if (!title) {
  //     throw new Error("No se envio el title");
  //   }

  return (
    <Fragment>
      {/* <code>{JSON.stringify(nombre)}</code> */}
      {/* <h1>{funcion()}</h1> */}

      <h1 data-testid="test-title"> {title} </h1>
      <p>{subtitulo}</p>
      <p>{subtitulo}</p>
    </Fragment>
  );
}

export default App;

App.propTypes = {
  title: PropTypes.string.isRequired, // PropTypes.string solo marcaria error que dice que se esperaba string, con .isRequired dice que tocaba mandar la prop
  subtitulo: PropTypes.string.isRequired,
};

// Las default props entran antes de las propTypes entonces si en propTypes hay algo requerido, pues utiliza el defaultProp. Para usar las defaultProps en el componente, tenemos que desestructuralas en los parametros
App.defaultProps = {
  title: "No hay titulo",
  subtitulo: "No hay subtitulo",
};
