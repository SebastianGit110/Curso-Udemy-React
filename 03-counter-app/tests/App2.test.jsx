import { render, screen } from "@testing-library/react";
import App from "../src/App";

describe("Pruebas en <App />", () => {
  const title = "Hola soy el titulo";
  const subTitle = "Hola soy el subtitulo";

  test("Debe hacer match con el snapshot", () => {
    const { container } = render(<App title={title} />);
    expect(container).toMatchSnapshot();
  });

  test("Debe mostrar el mensaje de title", () => {
    render(<App title={title} />); // No hay necesidad de hacer desestructuracion porque react-testing-library ofrece el screen que es el objeto que estamos renderizando
    // screen.debug(); // Imprime el componente renderizado
    expect(screen.getByText(title)).toBeTruthy(); // .not.toBeTruthy(); hace la negacion
  });

  test("Debe mostrar el title en un h1 no h2...", () => {
    render(<App title={title} />);
    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toContain(
      title
    ); // Esta prueba dice que el title debe estar en una etiqueta h1 no h2 ni h3
  });

  test("Debe mostrar el subtitulo enviado por props", () => {
    render(<App title={title} subtitulo={subTitle} />);
    expect(screen.getAllByText(subTitle).length).toBe(2);
  });
});
