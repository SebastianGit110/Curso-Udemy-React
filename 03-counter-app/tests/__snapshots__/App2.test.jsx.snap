// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <App /> Debe hacer match con el snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
     
    Hola soy el titulo
     
  </h1>
  <p>
    No hay subtitulo
  </p>
  <p>
    No hay subtitulo
  </p>
</div>
`;
